### DISCUSSION TITLE
Move HitObject.ApplyDefaultsToSelf effects to getters

### DISCUSSION THREAD

**Discussion by Hecatia-Lapislazuli** [ppy/osu#28402] [Ideas]
Currently, ApplyDefaultsToSelf can cause issues, at least in the editor, with certain behaviours becoming desynced. For example with what prompted this discussion, is a hit object may move to a different control point section, but behaviours would not update because nothing triggers a call to ApplyDefaults. I believe I've encountered similar behaviour when beatmap difficulty changes too.

I've checked through all implementations of this method, and from what I can tell, every variable it modifies bar TimePreempt should be easy to be moved to getters.

An exception does occur for hit windows, but this can most likely be left as is, or how hit windows works can be modified.

As for potential performance issues, given most calculations are just addition / multiplication, I think they should be negligible. The only non-addition / multiplication i can see is just finding the current control point which is done by a binary search, so should be optimized sufficiently regardless.

**Comment by peppy** [comment]
Move to which getters? To fix which issue?

**Reply by Hecatia-Lapislazuli** [reply]
I'm of the opinion all relevant getters, but specifically in this case for Slider, `Velocity` and `TickDistance`. https://github.com/ppy/osu/pull/28382#discussion_r1627372776 is what prompted this, but I see no real reason why this shouldn't also apply to other cases like for spinners with `SpinsRequired` and `MaximumBonusSpins`, or for juice streams with `velocityFactor` and `tickDistanceFactor` as examples.

**Reply by Hecatia-Lapislazuli** [reply]
Thinking about it, can I just do a PR of the proposed changes, regardless of whether the change would be accepted or not? 

Should only take an hour or two to do, so wouldn't be a big deal if they're not accepted. Plus I'd hope that would make it easier to say whether it's a good or bad change all around, having something demonstrable.

**Comment by bdach** [comment]
I will look at the PR that sparked this and give commentary there but this is 100% the wrong direction. I believe there's a chance you are just not doing things correctly (like attempting to mutate hitobjects while bypassing the editor beatmap change flow which would have applied defaults)

**Reply by Hecatia-Lapislazuli** [reply]
Hmm... fair point. I'm not exactly sure what i'd be doing wrong with editor beatmap change flow, but if there's something to learn in that regard for the future, I'd appreciate it.

I should be covered via a BeginChange and EndChange from here, but I'm not sure what else is needed, as in other situations this seemed sufficient.
https://github.com/ppy/osu/blob/318598730b0fea6c2fa66217a6c39b9998771486/osu.Game/Screens/Edit/Timing/TimingAdjustButton.cs#L78-L82

Although regardless of that, the issue existed before I did any work on the code, as demonstrable on my copy of lazer from 2023 (also works on master). After a BPM change, you have to independently do some action that will call ApplyDefaults for the slider length to update.

<img height="150" alt="image" src="https://github.com/ppy/osu/assets/7444073/7f4284ed-1330-462d-b2d5-1d9a9b233787">
<img height="150" alt="image" src="https://github.com/ppy/osu/assets/7444073/85669b6b-0323-47f2-98c6-7fba0043caec">
